From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: VerteDinde <khammond@slack-corp.com>
Date: Thu, 15 Jul 2021 12:16:50 -0700
Subject: chore: add gin::wrappable wrapperinfo crash key

This patch adds an additional crash key for gin::Wrappable, to help
debug a crash that is occurring during garbage collection in SecondWeakCallback.

The crash seems to be due to a class that is holding a reference to
gin::Wrappable even after being deleted. This added crash key compares
the soon-to-be-deleted WrapperInfo with known WrapperInfo components to
help determine where the crash originated from.

This patch should not be upstreamed, and can be removed in Electron 15 and
beyond once we identify the cause of the crash.

diff --git a/gin/wrappable.cc b/gin/wrappable.cc
index fe07eb94a8e679859bba6d76ff0d6ee86bd0c67e..d0066fca501eae5be4177440b44dbecc8e34c897 100644
--- a/gin/wrappable.cc
+++ b/gin/wrappable.cc
@@ -8,6 +8,10 @@
 #include "gin/object_template_builder.h"
 #include "gin/per_isolate_data.h"
 
+#if !defined(MAS_BUILD)
+#include "electron/shell/common/crash_keys.h"
+#endif
+
 namespace gin {
 
 WrappableBase::WrappableBase() = default;
@@ -36,6 +40,12 @@ void WrappableBase::FirstWeakCallback(
 void WrappableBase::SecondWeakCallback(
     const v8::WeakCallbackInfo<WrappableBase>& data) {
   WrappableBase* wrappable = data.GetParameter();
+
+#if !defined(MAS_BUILD)
+  WrapperInfo* info = static_cast<WrapperInfo*>(data.GetInternalField(0));
+  electron::crash_keys::SetCrashKeyForGinWrappable(info);
+#endif
+
   delete wrappable;
 }
 
